src/coprocessor/dag/expr/ctx.rs

```diff
@@ -47,6 +47,136 @@ pub const FLAG_DIVIDED_BY_ZERO_AS_WARNING: u64 = 1 << 8;
 pub const MODE_ERROR_FOR_DIVISION_BY_ZERO: u64 = 27;
 
 const DEFAULT_MAX_WARNING_CNT: usize = 64;
+
+pub struct EvalConfigBuilder {
+    /// timezone
+    tz: FixedOffset,
+    ignore_truncate: bool,
+    truncate_as_warning: bool,
+    overflow_as_warning: bool,
+    in_insert_stmt: bool,
+    in_update_or_delete_stmt: bool,
+    in_select_stmt: bool,
+    pad_char_to_full_length: bool,
+    divided_by_zero_as_warning: bool,
+    max_warning_cnt: usize,
+    sql_mode: u64,
+    strict_sql_mode: bool,
+}
+
+impl EvalConfigBuilder {
+    pub fn new() -> EvalConfigBuilder {
+        EvalConfigBuilder {
+            tz: FixedOffset::east(0),
+            ignore_truncate: false,
+            truncate_as_warning: false,
+            overflow_as_warning: false,
+            in_insert_stmt: false,
+            in_update_or_delete_stmt: false,
+            in_select_stmt: false,
+            pad_char_to_full_length: false,
```

Comment:
So many repeated attributes in `EvalConfigBuilder` and `EvalConfig`. And seems once we add a new attribute, we need to add it both in `EvalConfigBuilder` and `EvalConfig`. Maybe we could use flags instead, and set it in `EvalConfig` instead? It makes me confused, what's the benefit of `EvalConfigBuilder` @breeswish

Reply:
It is used to assign default values. Maybe we can just integrate it with the EvalConfig itself?

Reply:
IMO we can compress the config in `EvalConfig` as flags and make it immutable and remove the setters in `EvalConfig`. Will `EvalConfig` ever be modified in use cases?

Reply:
I mean, how about integrate `EvalConfigBuilder`'s set style into `EvalConfig` itself? You see, currently, we need to do this for eval config:

```rust
let config = EvalConfig::new(..);
config.set_xxx(xxx);
config.set_yyy(yyy);
```

So, with `EvalConfigBuilder`'s set style, we can chain them:

```rust
let config = EvalConfig::new(...)
    .set_xxx(xxx)
    .set_yyy(yyy);
```

With this, we can have both benefits of:

- no more repeated fields in two structs. there is only struct, EvalConfig
- we can use the lovely chaining syntax to easily set values as well as defaults.

Reply:
This seems reasonable @breeswish . I think builders come to be most useful when the `.build()` is failable or there is an expensive initialization.

Reply:
@ice1000 Take a look?

Reply:
Ok...

Reply:
Donr